src/main/java/com/uid2/admin/vertx/service/SharingService.java

```diff
@@ -15,82 +15,227 @@
 import java.util.stream.Collectors;
 
 public class SharingService implements IService {
     private final AuthMiddleware auth;
 
     private final WriteLock writeLock;
-    private final KeyAclStoreWriter storeWriter;
-    private final RotatingKeyAclProvider keyAclProvider;
+    private final KeysetStoreWriter storeWriter;
+    private final RotatingKeysetProvider keysetProvider;
+    private final IKeysetKeyManager keyManager;
     private static final Logger LOGGER = LoggerFactory.getLogger(SharingService.class);
 
     public SharingService(AuthMiddleware auth,
                           WriteLock writeLock,
-                          KeyAclStoreWriter storeWriter,
-                          RotatingKeyAclProvider keyAclProvider) {
+                          KeysetStoreWriter storeWriter,
+                          RotatingKeysetProvider keysetProvider,
+                          IKeysetKeyManager keyManager) {
         this.auth = auth;
         this.writeLock = writeLock;
         this.storeWriter = storeWriter;
-        this.keyAclProvider = keyAclProvider;
+        this.keysetProvider = keysetProvider;
+        this.keyManager = keyManager;
     }
 
     @Override
     public void setupRoutes(Router router) {
         router.get("/api/sharing/lists").handler(
-                auth.handle(this::handleKeyAclListAll, Role.SHARING_PORTAL)
+                auth.handle(this::handleAllowlistListALL, Role.SHARING_PORTAL)
         );
         router.get("/api/sharing/list/:siteId").handler(
-                auth.handle(this::handleKeyAclList, Role.SHARING_PORTAL)
+                auth.handle(this::handleAllowlistList, Role.SHARING_PORTAL)
         );
         router.post("/api/sharing/list/:siteId").handler(
-                auth.handle(this::handleKeyAclSet, Role.SHARING_PORTAL)
+                auth.handle(this::handleAllowlistSet, Role.SHARING_PORTAL)
         );
+
+        router.get("/api/sharing/keysets").handler(
+                auth.handle(this::handleKeysetListAll, Role.ADMINISTRATOR)
+        );
+        router.post("/api/sharing/keyset").handler(
+                auth.handle(this::handleKeysetSet, Role.ADMINISTRATOR)
+        );
+        router.get("/api/sharing/keyset/:keyset_id").handler(
+                auth.handle(this::handleKeysetList, Role.ADMINISTRATOR)
+        );
+    }
+
+    private void handleKeysetSet(RoutingContext rc) {
+        synchronized (writeLock) {
+           try {
+               keysetProvider.loadContent();
+           } catch (Exception e) {
+               LOGGER.error("Failed to load key acls");
+               rc.fail(500);
+           }
+
+            final JsonObject body = rc.body().asJsonObject();
+
+           final JsonArray whitelist = body.getJsonArray("allowlist");
+           Integer keysetId = body.getInteger("keyset_id");
+           final Integer siteId = body.getInteger("site_id");
+
+           final Map<Integer, Keyset> collection = this.keysetProvider.getSnapshot().getAllKeysets();
```

Comment:
The name is very generic and even slightly inaccurate. How about `keysetsById` or something similar?